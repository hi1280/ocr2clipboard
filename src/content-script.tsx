import { Close } from '@material-ui/icons';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import Measure from 'react-measure';
import styled from 'styled-components';
const Wrapper = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
`;

const Canvas = styled.canvas`
  position: fixed;
  top: 0;
  left: 0;
  cursor: crosshair;
`;

const TextArea = styled.div`
  position: absolute;
  background: #fff;
  border: 1px solid #000;
`;

const RightAlignDiv = styled.div`
  text-align: right;
`;

class Main extends React.Component {
  public state = {
    apiKey: '',
    drag: false,
    mouse: {
      x: 0,
      y: 0,
    },
    rectangle: {
      height: 0,
      left: 0,
      rgb: { red: '255', green: '0', blue: '0' },
      top: 0,
      width: 0,
    },
    textarea: {
      style: {
        display: 'none',
        left: 0,
        top: 0,
      },
      value: '',
    },
    wrapper: {
      height: 0,
      left: 0,
      top: 0,
      width: 0,
    },
  };
  public imageCanvasRef: React.RefObject<HTMLCanvasElement>;
  public textareaRef: React.RefObject<HTMLTextAreaElement>;

  constructor(props: object) {
    super(props);
    this.imageCanvasRef = React.createRef();
    this.textareaRef = React.createRef();
    chrome.storage.sync.get('apiKey', v => {
      this.setState({ apiKey: v.apiKey });
    });
  }

  public render() {
    return (
      <React.Fragment>
        <Measure
          offset={true}
          onResize={contentRect => {
            this.setState({
              wrapper: contentRect.offset,
            });
          }}
        >
          {({ measureRef }) => (
            <Wrapper ref={measureRef}>
              <Canvas
                width={this.state.wrapper.width}
                height={this.state.wrapper.height}
                onMouseDown={e => this.onMouseDown(e.clientX, e.clientY)}
                onMouseUp={e => this.onMouseUp(e.target as HTMLCanvasElement)}
                onMouseMove={e => {
                  if (this.state.drag) {
                    this.onMouseMove(e.target as HTMLCanvasElement, e.clientX, e.clientY);
                  }
                }}
              />
              <TextArea style={this.state.textarea.style} onClick={() => this.exit()}>
                <RightAlignDiv>
                  <Close style={{ fontSize: 18 }} />
                </RightAlignDiv>
                <textarea
                  ref={this.textareaRef}
                  value={this.state.textarea.value}
                  readOnly={true}
                  onClick={e => e.stopPropagation()}
                />
              </TextArea>
            </Wrapper>
          )}
        </Measure>
        <canvas ref={this.imageCanvasRef} width={this.state.rectangle.width} height={this.state.rectangle.height} />
      </React.Fragment>
    );
  }

  private onMouseDown(x: number, y: number) {
    const rectangle = this.state.rectangle;
    rectangle.left = x;
    rectangle.top = y;
    this.setState({
      drag: true,
      rectangle,
    });
  }

  private onMouseUp(canvasEl: HTMLCanvasElement) {
    const ctx = canvasEl.getContext('2d') as CanvasRenderingContext2D;
    this.setState({
      drag: false,
    });
    ctx.clearRect(0, 0, canvasEl.width, canvasEl.height);
    if (this.state.rectangle.width > 0 && this.state.rectangle.height > 0) {
      chrome.runtime.sendMessage({});
      this.addMessageListener();
    }
  }

  private addMessageListener() {
    const listener = (request: { base64Img: string }, _: any, sendResponse: (response: any) => void) => {
      const base64Img = request.base64Img;
      const canvas = this.imageCanvasRef.current as HTMLCanvasElement;
      const ctx = canvas.getContext('2d') as CanvasRenderingContext2D;
      const img = new Image();
      img.src = base64Img;
      img.addEventListener('load', async () => {
        ctx.drawImage(
          img,
          this.state.rectangle.left,
          this.state.rectangle.top,
          this.state.rectangle.width,
          this.state.rectangle.height,
          0,
          0,
          this.state.rectangle.width,
          this.state.rectangle.height,
        );
        const res = await fetch(`https://vision.googleapis.com/v1/images:annotate?key=${this.state.apiKey}`, {
          body: JSON.stringify({
            requests: [
              {
                features: [
                  {
                    type: 'TEXT_DETECTION',
                  },
                ],
                image: {
                  content: canvas.toDataURL('image/png').split(',')[1],
                },
              },
            ],
          }),
          method: 'POST',
        });
        const json = await res.json();
        if (json && json.responses && json.responses[0].textAnnotations) {
          const textarea = this.textareaRef.current as HTMLTextAreaElement;
          this.setState({
            textarea: {
              style: {
                display: 'block',
                left: `${this.state.mouse.x}px`,
                top: `${this.state.mouse.y}px`,
              },
              value: json.responses[0].textAnnotations[0].description,
            },
          });
          textarea.select();
        }
      });
      chrome.runtime.onMessage.removeListener(listener);
      sendResponse(true);
      return true;
    };
    chrome.runtime.onMessage.addListener(listener);
  }

  private onMouseMove(canvasEl: HTMLCanvasElement, x: number, y: number) {
    const ctx = canvasEl.getContext('2d') as CanvasRenderingContext2D;
    const rectangle = this.state.rectangle;
    rectangle.width = x - this.state.rectangle.left;
    rectangle.height = y - this.state.rectangle.top;
    this.setState({
      mouse: {
        x,
        y,
      },
      rectangle,
    });
    ctx.clearRect(0, 0, canvasEl.width, canvasEl.height);
    ctx.strokeStyle = `rgb(${this.state.rectangle.rgb.red}, ${this.state.rectangle.rgb.green}, ${
      this.state.rectangle.rgb.blue
    })`;
    ctx.strokeRect(
      this.state.rectangle.left,
      this.state.rectangle.top,
      this.state.rectangle.width,
      this.state.rectangle.height,
    );
  }

  private exit() {
    document.body.removeChild(app);
    chrome.storage.sync.set({ running: false });
  }
}

const app = document.createElement('div');
document.body.appendChild(app);
ReactDOM.render(<Main />, app);
