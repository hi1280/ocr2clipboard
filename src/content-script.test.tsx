import { mount } from 'enzyme';
import * as React from 'react';
import { Main } from './content-script';

test('Snapshot Testing at Initialize', () => {
  const main = mount(<Main />);
  expect(main).toMatchSnapshot();
});
