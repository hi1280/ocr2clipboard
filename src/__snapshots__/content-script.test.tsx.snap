// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing at Initialize 1`] = `
<Main>
  <Measure
    offset={true}
    onResize={[Function]}
  >
    <Component
      contentRect={
        Object {
          "bounds": Object {},
          "client": Object {},
          "entry": Object {},
          "margin": Object {},
          "offset": Object {},
          "scroll": Object {},
        }
      }
      measure={[Function]}
      measureRef={[Function]}
      offset={true}
    >
      <styled.div>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bdVaJa",
                "isStatic": true,
                "lastClassName": "dAWacL",
                "rules": Array [
                  "
  position: fixed;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-bdVaJa",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={[Function]}
        >
          <div
            className="sc-bdVaJa dAWacL"
          >
            <styled.canvas
              height={0}
              onMouseDown={[Function]}
              onMouseMove={[Function]}
              onMouseUp={[Function]}
              width={0}
            >
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bwzfXH",
                      "isStatic": true,
                      "lastClassName": "kNvjxv",
                      "rules": Array [
                        "
  position: fixed;
  top: 0;
  left: 0;
  cursor: crosshair;
",
                      ],
                    },
                    "displayName": "styled.canvas",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-bwzfXH",
                    "target": "canvas",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
                height={0}
                onMouseDown={[Function]}
                onMouseMove={[Function]}
                onMouseUp={[Function]}
                width={0}
              >
                <canvas
                  className="sc-bwzfXH kNvjxv"
                  height={0}
                  onMouseDown={[Function]}
                  onMouseMove={[Function]}
                  onMouseUp={[Function]}
                  width={0}
                />
              </StyledComponent>
            </styled.canvas>
            <styled.div
              onClick={[Function]}
              style={
                Object {
                  "display": "none",
                  "left": 0,
                  "top": 0,
                }
              }
            >
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-htpNat",
                      "isStatic": true,
                      "lastClassName": "kKJWOp",
                      "rules": Array [
                        "
  position: absolute;
  background: #fff;
  border: 1px solid #000;
",
                      ],
                    },
                    "displayName": "styled.div",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-htpNat",
                    "target": "div",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
                onClick={[Function]}
                style={
                  Object {
                    "display": "none",
                    "left": 0,
                    "top": 0,
                  }
                }
              >
                <div
                  className="sc-htpNat kKJWOp"
                  onClick={[Function]}
                  style={
                    Object {
                      "display": "none",
                      "left": 0,
                      "top": 0,
                    }
                  }
                >
                  <styled.div>
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-bxivhb",
                            "isStatic": true,
                            "lastClassName": "gJaspy",
                            "rules": Array [
                              "
  text-align: right;
",
                            ],
                          },
                          "displayName": "styled.div",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-bxivhb",
                          "target": "div",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                    >
                      <div
                        className="sc-bxivhb gJaspy"
                      >
                        <pure(CloseIcon)
                          style={
                            Object {
                              "fontSize": 18,
                            }
                          }
                        >
                          <CloseIcon
                            style={
                              Object {
                                "fontSize": 18,
                              }
                            }
                          >
                            <WithStyles(SvgIcon)
                              style={
                                Object {
                                  "fontSize": 18,
                                }
                              }
                            >
                              <SvgIcon
                                classes={
                                  Object {
                                    "colorAction": "MuiSvgIcon-colorAction-4",
                                    "colorDisabled": "MuiSvgIcon-colorDisabled-6",
                                    "colorError": "MuiSvgIcon-colorError-5",
                                    "colorPrimary": "MuiSvgIcon-colorPrimary-2",
                                    "colorSecondary": "MuiSvgIcon-colorSecondary-3",
                                    "fontSizeInherit": "MuiSvgIcon-fontSizeInherit-7",
                                    "fontSizeLarge": "MuiSvgIcon-fontSizeLarge-9",
                                    "fontSizeSmall": "MuiSvgIcon-fontSizeSmall-8",
                                    "root": "MuiSvgIcon-root-1",
                                  }
                                }
                                color="inherit"
                                component="svg"
                                fontSize="default"
                                style={
                                  Object {
                                    "fontSize": 18,
                                  }
                                }
                                viewBox="0 0 24 24"
                              >
                                <svg
                                  aria-hidden="true"
                                  className="MuiSvgIcon-root-1"
                                  focusable="false"
                                  role="presentation"
                                  style={
                                    Object {
                                      "fontSize": 18,
                                    }
                                  }
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                                  />
                                  <path
                                    d="M0 0h24v24H0z"
                                    fill="none"
                                  />
                                </svg>
                              </SvgIcon>
                            </WithStyles(SvgIcon)>
                          </CloseIcon>
                        </pure(CloseIcon)>
                      </div>
                    </StyledComponent>
                  </styled.div>
                  <textarea
                    onClick={[Function]}
                    readOnly={true}
                    value=""
                  />
                </div>
              </StyledComponent>
            </styled.div>
          </div>
        </StyledComponent>
      </styled.div>
    </Component>
  </Measure>
  <canvas
    height={0}
    width={0}
  />
</Main>
`;
